tests/test_swmm.cpp

```diff
@@ -0,0 +1,172 @@
+#define DATA_PATH_OUT "./swmm_api_test.out"
+
+#define DATA_PATH_INP2 "./swmm_api_test2.inp"
+#define DATA_PATH_RPT2 "./swmm_api_test2.rpt"
+#define DATA_PATH_OUT2 "./swmm_api_test2.out"
+
+using namespace std;
+
+// Custom test to check the minimum number of correct decimal digits between 
+// the test and the ref vectors.  
+bool tolCheck(float a, float b, float tol)
+{
+    return abs(a - b) <= tol;
+}
+
+BOOST_AUTO_TEST_SUITE (test_swmm_auto)
+
+BOOST_AUTO_TEST_CASE(RunTest) {
+    int error = swmm_run( (char *)DATA_PATH_INP, (char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT);
+    BOOST_CHECK_EQUAL(0, error);
+}
+
+BOOST_AUTO_TEST_CASE(OpenTest) {
+    int error = swmm_open((char *)DATA_PATH_INP, (char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT);
+    BOOST_CHECK_EQUAL(0, error);
+    swmm_close();
+}
+
+// Try to open valid file but not an input file - Should Fail at 200?
+BOOST_AUTO_TEST_CASE(FailOpen191) {
+    int error = swmm_open((char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT, (char *)"");
+    BOOST_CHECK_EQUAL(191, error);
+    swmm_close();
+}
+
+BOOST_AUTO_TEST_CASE(FailOpen301) {
+    int error = swmm_open((char *)DATA_PATH_INP, (char *)DATA_PATH_INP, (char *)DATA_PATH_INP);
+    BOOST_CHECK_EQUAL(301, error);
+    swmm_close();
+}
+
+BOOST_AUTO_TEST_CASE(FailOpen303) {
+    int error = swmm_open((char *)"", (char *)DATA_PATH_INP, (char *)DATA_PATH_OUT);
+    BOOST_CHECK_EQUAL(303, error);
+    swmm_close();
+}
+
+BOOST_AUTO_TEST_CASE(FailOpen305) {
+    int error = swmm_open((char *)DATA_PATH_INP, (char *)"", (char *)DATA_PATH_OUT);
+    BOOST_CHECK_EQUAL(305, error);
+    swmm_close();
+}
+
+BOOST_AUTO_TEST_CASE(FailOpen307) {
+    FILE * fp;
+    fp = fopen(DATA_PATH_OUT, "r");
+    int error = swmm_open((char *)DATA_PATH_INP, (char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT);
+    BOOST_CHECK_EQUAL(0, error);// Should be 307
+    swmm_close();
+    fclose(fp);
+}
+
+BOOST_AUTO_TEST_SUITE_END()
+
+// Adding Fixtures
+struct FixtureStep{
+    FixtureStep() {
+        swmm_open((char *)DATA_PATH_INP, (char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT);
+    }
+    ~FixtureStep() {
+        swmm_end();
+        swmm_close();
+    }
+};
+
+struct FixtureBeforeClose{
+    FixtureBeforeClose() {
+        swmm_open((char *)DATA_PATH_INP, (char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT);
+        swmm_start(0);
+
+        int error;
+        double elapsedTime = 0.0;
+        do
+        {
+            error = swmm_step(&elapsedTime);
+        }while (elapsedTime != 0 && !error);
+        BOOST_CHECK_EQUAL(0, error);
+        swmm_end();
+    }
+    ~FixtureBeforeClose() {
+        swmm_close();
+    }
+};
+
+struct FixtureBeforeStart{
+    FixtureBeforeStart() {
+        swmm_open((char *)DATA_PATH_INP, (char *)DATA_PATH_RPT, (char *)DATA_PATH_OUT);
+    }
+    ~FixtureBeforeStart() {
+        int error;
+        double elapsedTime = 0.0;
+        do
+        {
+            error = swmm_step(&elapsedTime);
+        }while (elapsedTime != 0 && !error);
```

Comment:
Why are you running the simulation in the fixture destructor? I suppose you have a good reason for doing it, but it seems odd.

Reply:
@michaeltryby, the idea of these fixtures are to isolate different parts of the simulation state.  Certain things work at certain times.